roles/st2chatops/tasks/main.yml

```diff
@@ -26,18 +27,55 @@
   when: st2chatops_version != "latest"
   tags: [st2chatops, skip_ansible_lint]
 
-- name: Add st2_api_key in st2chatops.env
+- name: Check if api key already exist. No change if it already exists
+  become: yes
+  shell: "grep -e '^export ST2_API_KEY=\"\\${ST2_API_KEY}\"$' /opt/stackstorm/chatops/st2chatops.env"
+  changed_when: no
+  ignore_errors: True
+  register: task_apikey_exists
+
+- name: Add user defined "st2chatops_st2_api_key" in st2chatops.env, from "role" or "defaults/main.yml"
   become: yes
   replace:
     dest: /opt/stackstorm/chatops/st2chatops.env
-    regexp: '(?<=ST2_API_KEY=)(.*)$'
+    regexp: '(?<=ST2_API_KEY=)(\"\$\{ST2_API_KEY\}\")$'
     replace: '{{ st2chatops_st2_api_key }}'
-  when: st2chatops_st2_api_key != 'CHANGE-ME-PLEASE'
-  register: task_st2_api_key
+  when: >
+    (task_apikey_exists.failed is not defined) and
+    (st2chatops_st2_api_key is defined) and
+    (st2chatops_st2_api_key != "CHANGE-ME-PLEASE")
+  register: task_user_st2_api_key
   no_log: true
   notify: restart st2chatops
   tags: st2chatops
 
+- name: Generate authentication token
+  command: st2 auth {{ st2_auth_username }} -p {{ st2_auth_password }} -t
+  when: (task_apikey_exists.failed is not defined) and (task_user_st2_api_key.changed == false)
+  register: task_st2_token
+  tags: [st2chatops, skip_ansible_lint]
+
+- name: Generate "st2_api_key" if not provided with the "role" or in "defaults/main.yml"
+  command: st2 apikey create -k
+  environment:
+    ST2_AUTH_TOKEN: "{{ task_st2_token.stdout }}"
+  when: task_st2_token.changed
+  register: task_generated_api_key
+  no_log: true
+  tags: [st2chatops, skip_ansible_lint]
+
+- name: Add generated "st2_api_key" in st2chatops.env
+  become: yes
+  replace:
```

Comment:
@armab: AFAIU, first "set API key" is for `Add user defined "st2chatops_st2_api_key" in st2chatops.env, from "role" or "defaults/main.yml"` and second for `Add generated "st2_api_key" in st2chatops.env` with different input and conditions. Also, we don't want to generate api key every time we execute this task. The only way that i can think of is moving this task in a separate yaml file.